File: backend/app/api/v1/endpoints/onboarding.py
"""
Onboarding API Endpoints
"""

from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel

from app.services.onboarding_service import OnboardingService
from app.models.user import User
from app.models.onboarding import OnboardingStep
from app.dependencies import get_current_user
from app.core.database import get_db
from sqlalchemy.ext.asyncio import AsyncSession

router = APIRouter()


class OnboardingStepResponse(BaseModel):
    id: int
    key: str
    title: str
    description: Optional[str]
    order: int
    step_type: str
    step_data: Optional[dict]
    required: bool

    class Config:
        from_attributes = True


class OnboardingProgressResponse(BaseModel):
    is_completed: bool
    current_step: Optional[str]
    completed_count: int
    total_count: int
    progress_percentage: float


@router.get("/onboarding/steps", response_model=List[OnboardingStepResponse], tags=["onboarding"])
async def get_onboarding_steps(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get active onboarding steps for the current user"""
    from app.services.rbac_service import RBACService
    
    service = OnboardingService(db)
    rbac_service = RBACService(db)
    user_roles = await rbac_service.get_user_roles(current_user.id)
    # Convert roles to list of role slugs for the service
    user_role_slugs = [role.slug for role in user_roles] if user_roles else None
    steps = await service.get_active_steps(user_roles=user_role_slugs)
    return [OnboardingStepResponse.model_validate(s) for s in steps]


@router.get("/onboarding/progress", response_model=OnboardingProgressResponse, tags=["onboarding"])
async def get_onboarding_progress(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get onboarding progress for the current user"""
    service = OnboardingService(db)
    progress = await service.get_progress(current_user.id)
    return OnboardingProgressResponse(**progress)


@router.get("/onboarding/next-step", response_model=Optional[OnboardingStepResponse], tags=["onboarding"])
async def get_next_step(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get the next onboarding step for the current user"""
    service = OnboardingService(db)
    # Initialize if needed
    await service.initialize_onboarding(current_user.id)
    step = await service.get_next_step(current_user.id)
    if step:
        return OnboardingStepResponse.model_validate(step)
    return None


@router.post("/onboarding/initialize", tags=["onboarding"])
async def initialize_onboarding(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Initialize onboarding for the current user"""
    from app.services.rbac_service import RBACService
    
    service = OnboardingService(db)
    rbac_service = RBACService(db)
    user_roles = await rbac_service.get_user_roles(current_user.id)
    # Convert roles to list of role slugs for the service
    user_role_slugs = [role.slug for role in user_roles] if user_roles else None
    onboarding = await service.initialize_onboarding(current_user.id, user_roles=user_role_slugs)
    return {"success": True, "message": "Onboarding initialized"}


@router.post("/onboarding/steps/{step_key}/complete", tags=["onboarding"])
async def complete_step(
    step_key: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Mark an onboarding step as completed"""
    service = OnboardingService(db)
    onboarding = await service.complete_step(current_user.id, step_key)
    if not onboarding:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Onboarding not initialized"
        )
    return {"success": True, "message": "Step completed"}


@router.post("/onboarding/steps/{step_key}/skip", tags=["onboarding"])
async def skip_step(
    step_key: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Skip an onboarding step"""
    service = OnboardingService(db)
    onboarding = await service.skip_step(current_user.id, step_key)
    if not onboarding:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Onboarding not initialized"
        )
    return {"success": True, "message": "Step skipped"}


# Employee onboarding endpoints
@router.get("/employees/{employee_id}/onboarding/progress", response_model=OnboardingProgressResponse, tags=["onboarding", "employees"])
async def get_employee_onboarding_progress(
    employee_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get onboarding progress for an employee"""
    from app.models.employee import Employee
    from sqlalchemy import select
    
    # Get employee
    result = await db.execute(select(Employee).where(Employee.id == employee_id))
    employee = result.scalar_one_or_none()
    
    if not employee:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Employee not found"
        )
    
    if not employee.user_id:
        # Employee doesn't have a user account yet
        return {
            "is_completed": False,
            "current_step": None,
            "completed_count": 0,
            "total_count": 0,
            "progress_percentage": 0.0
        }
    
    service = OnboardingService(db)
    progress = await service.get_progress(employee.user_id)
    return OnboardingProgressResponse(**progress)


@router.get("/employees/{employee_id}/onboarding/steps", response_model=List[OnboardingStepResponse], tags=["onboarding", "employees"])
async def get_employee_onboarding_steps(
    employee_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get onboarding steps for an employee"""
    from app.models.employee import Employee
    from app.services.rbac_service import RBACService
    from sqlalchemy import select
    
    # Get employee
    result = await db.execute(select(Employee).where(Employee.id == employee_id))
    employee = result.scalar_one_or_none()
    
    if not employee:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Employee not found"
        )
    
    service = OnboardingService(db)
    
    # Get user roles if employee has a user account
    user_role_slugs = None
    if employee.user_id:
        rbac_service = RBACService(db)
        user_roles = await rbac_service.get_user_roles(employee.user_id)
        user_role_slugs = [role.slug for role in user_roles] if user_roles else None
    
    steps = await service.get_active_steps(user_roles=user_role_slugs)
    return [OnboardingStepResponse.model_validate(s) for s in steps]


@router.post("/employees/{employee_id}/onboarding/initialize", tags=["onboarding", "employees"])
async def initialize_employee_onboarding(
    employee_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Initialize onboarding for an employee"""
    from app.models.employee import Employee
    from app.services.rbac_service import RBACService
    from sqlalchemy import select
    
    # Get employee
    result = await db.execute(select(Employee).where(Employee.id == employee_id))
    employee = result.scalar_one_or_none()
    
    if not employee:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Employee not found"
        )
    
    if not employee.user_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Employee must have a user account to initialize onboarding"
        )
    
    service = OnboardingService(db)
    rbac_service = RBACService(db)
    user_roles = await rbac_service.get_user_roles(employee.user_id)
    user_role_slugs = [role.slug for role in user_roles] if user_roles else None
    onboarding = await service.initialize_onboarding(employee.user_id, user_roles=user_role_slugs)
    return {"success": True, "message": "Onboarding initialized"}


@router.post("/employees/{employee_id}/onboarding/steps/{step_key}/complete", tags=["onboarding", "employees"])
async def complete_employee_onboarding_step(
    employee_id: int,
    step_key: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Mark an onboarding step as completed for an employee"""
    from app.models.employee import Employee
    from sqlalchemy import select
    
    # Get employee
    result = await db.execute(select(Employee).where(Employee.id == employee_id))
    employee = result.scalar_one_or_none()
    
    if not employee:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Employee not found"
        )
    
    if not employee.user_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Employee must have a user account"
        )
    
    service = OnboardingService(db)
    onboarding = await service.complete_step(employee.user_id, step_key)
    if not onboarding:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Onboarding not initialized"
        )
    return {"success": True, "message": "Step completed"}


@router.get("/employees/onboarding/list", tags=["onboarding", "employees"])
async def list_employees_onboarding(
    team_id: Optional[int] = None,
    skip: int = 0,
    limit: int = 1000,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """List all employees with their onboarding status"""
    from app.models.employee import Employee
    from sqlalchemy import select, and_
    
    query = select(Employee)
    if team_id:
        query = query.where(Employee.team_id == team_id)
    
    query = query.offset(skip).limit(limit)
    result = await db.execute(query)
    employees = result.scalars().all()
    
    service = OnboardingService(db)
    onboarding_list = []
    
    for employee in employees:
        progress_data = {
            "is_completed": False,
            "current_step": None,
            "completed_count": 0,
            "total_count": 0,
            "progress_percentage": 0.0
        }
        
        if employee.user_id:
            progress_data = await service.get_progress(employee.user_id)
        
        onboarding_list.append({
            "employee_id": employee.id,
            "employee_name": f"{employee.first_name} {employee.last_name}",
            "employee_email": employee.email,
            "hire_date": employee.hire_date.isoformat() if employee.hire_date else None,
            "team_id": employee.team_id,
            "user_id": employee.user_id,
            "progress": progress_data
        })
    
    return onboarding_list

File: apps/web/src/lib/api/onboarding.ts
/**
 * Onboarding API
 * API client for onboarding endpoints
 */

import { apiClient } from './client';
import { extractApiData } from './utils';

export interface OnboardingStep {
  id: number;
  key: string;
  title: string;
  description?: string | null;
  order: number;
  step_type: string;
  step_data?: Record<string, unknown> | null;
  required: boolean;
}

export interface OnboardingProgress {
  is_completed: boolean;
  current_step: string | null;
  completed_count: number;
  total_count: number;
  progress_percentage: number;
}

export interface EmployeeOnboardingItem {
  employee_id: number;
  employee_name: string;
  employee_email?: string | null;
  hire_date?: string | null;
  team_id?: number | null;
  user_id?: number | null;
  progress: OnboardingProgress;
}

/**
 * Onboarding API client
 */
export const onboardingAPI = {
  /**
   * Get onboarding steps for current user
   */
  getSteps: async (): Promise<OnboardingStep[]> => {
    const response = await apiClient.get<OnboardingStep[]>('/v1/onboarding/steps');
    return extractApiData<OnboardingStep[]>(response);
  },

  /**
   * Get onboarding progress for current user
   */
  getProgress: async (): Promise<OnboardingProgress> => {
    const response = await apiClient.get<OnboardingProgress>('/v1/onboarding/progress');
    return extractApiData<OnboardingProgress>(response);
  },

  /**
   * Initialize onboarding for current user
   */
  initialize: async (): Promise<{ success: boolean; message: string }> => {
    const response = await apiClient.post<{ success: boolean; message: string }>('/v1/onboarding/initialize');
    return extractApiData<{ success: boolean; message: string }>(response);
  },

  /**
   * Complete an onboarding step for current user
   */
  completeStep: async (stepKey: string): Promise<{ success: boolean; message: string }> => {
    const response = await apiClient.post<{ success: boolean; message: string }>(
      `/v1/onboarding/steps/${stepKey}/complete`
    );
    return extractApiData<{ success: boolean; message: string }>(response);
  },

  /**
   * Skip an onboarding step for current user
   */
  skipStep: async (stepKey: string): Promise<{ success: boolean; message: string }> => {
    const response = await apiClient.post<{ success: boolean; message: string }>(
      `/v1/onboarding/steps/${stepKey}/skip`
    );
    return extractApiData<{ success: boolean; message: string }>(response);
  },

  /**
   * Get onboarding progress for an employee
   */
  getEmployeeProgress: async (employeeId: number): Promise<OnboardingProgress> => {
    const response = await apiClient.get<OnboardingProgress>(
      `/v1/employees/${employeeId}/onboarding/progress`
    );
    return extractApiData<OnboardingProgress>(response);
  },

  /**
   * Get onboarding steps for an employee
   */
  getEmployeeSteps: async (employeeId: number): Promise<OnboardingStep[]> => {
    const response = await apiClient.get<OnboardingStep[]>(
      `/v1/employees/${employeeId}/onboarding/steps`
    );
    return extractApiData<OnboardingStep[]>(response);
  },

  /**
   * Initialize onboarding for an employee
   */
  initializeEmployee: async (employeeId: number): Promise<{ success: boolean; message: string }> => {
    const response = await apiClient.post<{ success: boolean; message: string }>(
      `/v1/employees/${employeeId}/onboarding/initialize`
    );
    return extractApiData<{ success: boolean; message: string }>(response);
  },

  /**
   * Complete an onboarding step for an employee
   */
  completeEmployeeStep: async (
    employeeId: number,
    stepKey: string
  ): Promise<{ success: boolean; message: string }> => {
    const response = await apiClient.post<{ success: boolean; message: string }>(
      `/v1/employees/${employeeId}/onboarding/steps/${stepKey}/complete`
    );
    return extractApiData<{ success: boolean; message: string }>(response);
  },

  /**
   * List all employees with their onboarding status
   */
  listEmployeesOnboarding: async (options?: {
    team_id?: number;
    skip?: number;
    limit?: number;
  }): Promise<EmployeeOnboardingItem[]> => {
    const response = await apiClient.get<EmployeeOnboardingItem[]>('/v1/employees/onboarding/list', {
      params: {
        team_id: options?.team_id,
        skip: options?.skip || 0,
        limit: options?.limit || 1000,
      },
    });
    return extractApiData<EmployeeOnboardingItem[]>(response);
  },
};
